apps/dcos_dns/src/dcos_dns_handler.erl

```diff
@@ -379,4 +403,22 @@ init_metrics() ->
         {labels, [zone]},
         {duration_unit, seconds},
         {buckets, [0.001, 0.005, 0.010, 0.050, 0.100, 0.500, 1.000, 5.000]},
-        {help, "The time spent processing DNS requests."}]).
+        {help, "The time spent processing DNS requests."}]),
+    prometheus_counter:new([
+        {registry, dns},
+        {name, worker_failures_total},
+        {labels, [upstream_address, protocol, reason]},
+        {help, "Total number of worker failures processing DNS requests."}]),
+    prometheus_histogram:new([
+        {registry, dns},
+        {name, worker_requests_duration_seconds},
+        {labels, [upstream_address, protocol]},
+        {duration_unit, seconds},
+        {buckets, [0.001, 0.005, 0.010, 0.050, 0.100, 0.500, 1.000, 5.000]},
+        {help, "The time a worker spent processing DNS requests."}]).
+
+-spec(emit_worker_failure(binary(), udp | tcp, atom()) -> ok).
```

Comment:
Could you please try to move the logic from receive functions to parent functions and maybe in that case we don't have to introduce this one?